<a href="https://colab.research.google.com/github/Maheshcheegiti/dwdm-lab/blob/main/DWDM_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# connecting to google drive for datasets
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


###Experiment on building a model using linear regression algorithm on any dataset.
>“Linear regression” is a statistical method for predicting the value of a dependent variable based on one or more independent variables. It assumes that there is a linear relationship between the variables, and the goal is to find the line (or plane) that best fits the data and minimizes the difference between the predicted values and the actual values. Linear regression is commonly used for tasks like forecasting and trend analysis in various fields.

>The equation of the line in a simple linear regression with one independent variable is y = mx + b, where y is the dependent variable, x is the independent variable, m is the slope of the line, and b is the intercept. The values of m and b are determined by the linear regression algorithm using the least squares method

>Linear regression is commonly used in various fields, such as finance, economics, social sciences, engineering, and machine learning, for tasks such as forecasting, trend analysis, and causal inference.

>Dataset: "Boston Housing" dataset from Scikit-learn library

>Step 1: Loading the dataset
To load the Boston Housing dataset, we first need to import the required libraries and load the dataset as follows:


In [ ]:
import pandas as pd

# load the dataset
boston = pd.read_csv("/content/drive/MyDrive/datasets-master/BostonHousing.csv")

# convert it into a pandas dataframe
data = pd.DataFrame(boston)

data.head()

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

        b  lstat  medv  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2

> Step 2: Splitting the dataset into training and testing sets
We need to split the dataset into training and testing sets to evaluate the performance of the linear regression model. We can split the dataset using the train_test_split() function from the Scikit-learn library as follows:



In [ ]:
from sklearn.model_selection import train_test_split

# 'medv' is your target variable, and you want to predict it using other features
X = data.drop('medv', axis=1)  # Features
y = data['medv']  # Target variable

# Split the dataset into training and testing sets (e.g., 80% training, 20% testing)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


>Step 3: Building the linear regression model

>We can build the linear regression model using the LinearRegression() class from the Scikit-learn library as follows:


In [ ]:
from sklearn.linear_model import LinearRegression

# Create an instance of the LinearRegression model
linear_reg_model = LinearRegression()

# Fit the model to the training data
linear_reg_model.fit(X_train, y_train)

LinearRegression()

>Step 4: Evaluating the performance of the model
We can evaluate the performance of the linear regression model using the mean squared error (MSE) and R-squared (R^2) metrics. We can calculate these metrics using the mean_squared_error() and r2_score() functions from the Scikit-learn library as follows:


In [ ]:
from sklearn.metrics import mean_squared_error, r2_score

# Make predictions on the testing set
y_pred = linear_reg_model.predict(X_test)

# Calculate mean squared error (MSE)
mse = mean_squared_error(y_test, y_pred)
print("Mean Squared Error:", mse)

# Calculate R-squared (R^2) score
r2 = r2_score(y_test, y_pred)
print("R-squared (R^2) Score:", r2)

Mean Squared Error: 24.291119474973478
R-squared (R^2) Score: 0.6687594935356326


>Step 5: Making predictions on new data
Finally, we can use the trained linear regression model to make predictions on new data. We can make predictions using the Predict() function of the linear regression object as follows:


In [ ]:
import numpy as np

# create new data for prediction
new_data = np.array([[0.02, 20.0, 6.0, 0.0, 0.5, 6.5, 65.0, 4.0, 1.0, 296.0, 15.0, 396.9, 4.98]])

# make predictions on new data
new_pred = linear_reg_model.predict(new_data)

print("Predicted Value: ", new_pred)

Predicted Value:  [30.88806702]


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


###Output:
>Mean squared error: 24.291119474973616

>Coefficient of determination (R-squared): 0.6684825753971583

Here, we can see that the mean squared error is 24.29, which indicates that the model's predictions are off by an average of 24.29 units. The coefficient of determination (R-squared) is 0.668, which indicates that the model explains 66.8% of the variance in the target variable. This is a decent result, but there is certainly room for improvement. We could try feature engineering, regularization, or using a more complex model to see if we can improve the performance
